[C_API] Pass compilation in current state

xolott opened the pull request:
The current version of Faiss has c_api compiling issues due to the last v1.6.0 changes.

This pull request only changes the necessary to compile without problem.

Enet4 commented on c_api/index_io_c.cpp:
```diff
@@ -15,35 +15,40 @@
-    } CATCH_AND_HANDLE
-}
-
-int faiss_clone_index (const FaissIndex *idx, FaissIndex **p_out) {
```
`clone_index` still appears to be in the library, declared in [clone_index.h](https://github.com/facebookresearch/faiss/blob/master/clone_index.h).

xolott commented:
@Enet4 I'm using visual studio code with some auto-formatting extension. Sorry about that, I will revert the formating to align with the rest of the c_api.
 I have a question (I'm not a C/C++ expert): Why in every ###_new function is expected to receive a double pointer to `Index` and not a single pointer.

Enet4 commented:
> I have a question (I'm not a C/C++ expert): Why in every ###_new function is expected to receive a double pointer to `Index` and not a single pointer.

It is indeed a double pointer, to serve as an output parameter to a function which provides an extra "return" value. It would be used like this:

```c
FaissIndex* index = NULL;
int e = FlatIndex_new(&index, other_args); // passed a FaissIndex**
```

In other words, we're passing a place for the function to write a pointer to a newly created instance. How the actual instance is allocated is left as an implementation detail. Were there a single indirection point, the pattern would have to be different: the consumer would have to allocate space for the index and the function would need to construct the instance on that space. For the Faiss indexes, I felt the former approach to be more intuitive.

xolott commented:
This is the list of changes:

- Fixed formatting issues.
- Based on [this log](https://github.com/facebookresearch/faiss/pull/943#issuecomment-532590627):
```
index_io_c.h: split off clone_index_c.h
AuxIndexStructures_c.h: moved to impl/AuxIndexStructures_c.h
AutoTune_c.h: split off index_factory_c.h
```
- Updated Makefile with new files and modified paths
- Updated `example_c.c` to include `clone_index_c.h`
- Included `faiss_Index_compute_residual_n` to `Index_c.h`

Enet4 commented on c_api/impl/AuxIndexStructures_c.h:
```diff
@@ -126,8 +126,24 @@ int faiss_RangeSearchPartialResult_set_lims(
     FaissRangeSearchPartialResult* res, idx_t qno, FaissRangeQueryResult** qr);
 
+
+FAISS_DECLARE_CLASS(DistanceComputer)
```
The respective implementations of the functions below appear to be missing. Please follow the pattern seen in the other opaque types to fill in the missing definitions. You can also copy the rest of the documentation to each function where applicable.

xolott replied:
I just finish the definitions. But I have a trouble with `DistanceComputer::operator()`... I changed the name to `faiss_DistanceComputer_vector_to_query_dis`, but I want to read your suggestions about a better name.

Enet4 replied:
Well, I don't have a strong opinion regarding the name of the operator, but it's worth documenting that this function defers to the overloaded function call operator, so that users know where it comes from in the native API.

Enet4 commented on c_api/impl/AuxIndexStructures_c.h:
```diff
@@ -126,8 +126,24 @@ int faiss_RangeSearchPartialResult_set_lims(
+ * @param dc          opaque pointer to distance computer object
+ * @param key         output distance
+ **/
+int faiss_DistanceComputer_symmetrik_dis(FaissDistanceComputer *dc, idx_t i, idx_t j, float *vd);
```
It's `symmetric_dis` in the main API, it might be better to stick to the same name.

```suggestion
int faiss_DistanceComputer_symmetric_dis(FaissDistanceComputer *dc, idx_t i, idx_t j, float *vd);
```

xolott replied:
My bad... a typo

Enet4 commented on c_api/impl/AuxIndexStructures_c.h:
```diff
@@ -126,8 +126,24 @@ int faiss_RangeSearchPartialResult_set_lims(
+ * Compute distance between two stored vectors
+ *
+ * @param dc          opaque pointer to distance computer object
+ * @param key         output distance
```
It's named `vd` below.

```suggestion
 * @param vd          output distance
```

xolott commented:
All changes are already in the branch. If there is no other observation, this could be merge :smiley:

Enet4 commented on c_api/impl/AuxIndexStructures_c.h:
```diff
@@ -126,8 +126,24 @@ int faiss_RangeSearchPartialResult_set_lims(
+
+/**
+ * Compute distance of vector i to current query.
+ * This function correspond to the call operator DistanceComputer::operator()()
```
```suggestion
 * This function corresponds to the function call operator: DistanceComputer::operator()
```

Enet4 commented on c_api/impl/AuxIndexStructures_c.h:
```diff
@@ -126,8 +126,24 @@ int faiss_RangeSearchPartialResult_set_lims(
+/**
+ * Compute distance of vector i to current query.
+ * This function correspond to the call operator DistanceComputer::operator()()
+ * */
```
```suggestion
 */
```

xolott commented:
Done!

xolott commented:
No problem. Metrics updated

mdouze commented:
Merging, since @Enet4 approves. 
Sorry for the lag.
